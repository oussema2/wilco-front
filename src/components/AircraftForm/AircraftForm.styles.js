import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create( {
	buttonContainer: {
		alignSelf: 'center',
		marginVertical: 10
	},
	makeAndModelTextInput: {
		marginTop: 40,
		marginBottom: 24
	},
	tailNumberTextInput: {
		marginBottom: 36
	}
} );
