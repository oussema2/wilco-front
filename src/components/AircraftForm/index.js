export { default as AircraftForm } from './AircraftForm';
