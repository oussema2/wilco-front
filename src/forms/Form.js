import MobxReactForm from 'mobx-react-form';
import dvr from '../plugins/dvr';

export default class Form extends MobxReactForm {
	constructor(
		{ fields = [] } = {},
		{
			hooks,
			options = {
				validateOnChange: true,
				showErrorsOnChange: false,
				validationDebounceWait: 0
			},
			plugins = dvr
		} = {}
	) {
		super( { fields }, { options, hooks, plugins } );
	}
}
