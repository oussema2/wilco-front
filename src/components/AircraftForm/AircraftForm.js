/* eslint-disable react/jsx-no-bind */
import React from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react-lite';
import { View, Keyboard } from 'react-native';
import { NavigationBar } from '../NavigationBar';
import { EditAvatar } from '../EditAvatar';
import { PrimaryButton, TextInput } from '..';
import { HorizontalPadding } from '../HorizontalPadding';
import { MAKE_AND_MODEL, TAIL_NUMBER } from '../../constants/formFields/aircraftForm';
import { styles } from './AircraftForm.styles';
import Form from '../../forms/Form';
import { KeyboardAwareScrollView } from '../KeyboardAwareScrollView';

/* eslint-disable arrow-body-style */

const AircraftForm = ( { presenter } ) => {
	const _onSubmit = ( event ) => {
		Keyboard.dismiss();
		presenter.form.onSubmit( event );
	};

	const _fixedBottomContent = () => (
		<View style={styles.buttonContainer}>
			<PrimaryButton
				testID="submit-button"
				title={presenter.submitButtonTitle}
				size="small"
				disabled={presenter.isSubmitButtonDisabled}
				onPress={_onSubmit}
			/>
		</View>
	);

	return !!presenter && (
		<>
			<HorizontalPadding>
				<NavigationBar
					title={presenter.title}
					onBackArrowPress={presenter.backButtonWasPressed}
				/>
			</HorizontalPadding>

			<KeyboardAwareScrollView fixedBottomContent={_fixedBottomContent()}>
				<HorizontalPadding>
					<EditAvatar
						testID="edit-avatar"
						variant="aircraft"
						size="big"
						onAvatarChange={presenter.onAvatarChange}
						source={presenter.avatarSource}
					/>
					<TextInput
						testID="make-and-model-input"
						required
						autoCapitalize="words"
						textInputStyle={styles.makeAndModelTextInput}
						error={presenter.form.$( MAKE_AND_MODEL ).error}
						inputProps={presenter.form.$( MAKE_AND_MODEL ).bind()}
					/>
					<TextInput
						testID="tail-number-input"
						textInputStyle={styles.tailNumberTextInput}
						autoCapitalize="characters"
						inputProps={presenter.form.$( TAIL_NUMBER ).bind()}
						helperText="This information will not be visible to others."
					/>
				</HorizontalPadding>
			</KeyboardAwareScrollView>
		</>
	);
};

AircraftForm.propTypes = {
	presenter: PropTypes.shape( {
		title: PropTypes.string.isRequired,
		submitButtonTitle: PropTypes.string.isRequired,
		isSubmitButtonDisabled: PropTypes.bool,
		backButtonWasPressed: PropTypes.func.isRequired,
		onAvatarChange: PropTypes.func.isRequired,
		avatarSource: PropTypes.shape( {
			uri: PropTypes.string.isRequired
		} ),
		form: PropTypes.instanceOf( Form ).isRequired
	} )
};

AircraftForm.defaultProps = {
	presenter: null
};

export default observer( AircraftForm );
