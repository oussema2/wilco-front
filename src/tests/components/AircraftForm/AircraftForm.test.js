import React from 'react';
import { Keyboard } from 'react-native';
import { fireEvent, render, act } from '@testing-library/react-native';
import AircraftForm from '../../../components/AircraftForm/AircraftForm';
import Form from '../../../forms/Form';
import fields from '../../../forms/aircraftFields';

describe( 'AircraftForm', () => {
	let presenter;
	let component;
	const title = 'Do something with aircraft';
	const submitButtonTitle = 'Submit';
	const isSubmitButtonDisabled = false;
	const backButtonWasPressed = jest.fn();
	const onAvatarChange = jest.fn();
	const avatarSource = { uri: 'avatar/source' };
	const form = new Form( { fields } );

	const setUp = ( presenterProps = {} ) => {
		presenter = {
			title,
			submitButtonTitle,
			isSubmitButtonDisabled,
			form,
			onAvatarChange,
			avatarSource,
			backButtonWasPressed,
			...presenterProps
		};
		component = render( <AircraftForm presenter={presenter} /> );
	};

	beforeEach( () => {
		setUp();
	} );

	it( 'renders the AircraftForm screen correctly', () => {
		expect( component.getByText( title ) ).toBeDefined();
		expect( component.getByText( submitButtonTitle ) ).toBeDefined();
		expect( component.getByText( 'Change aircraft photo' ) ).toBeDefined();
		expect( component.getByText( 'Make and model' ) ).toBeDefined();
		expect( component.getByText( 'Tail number' ) ).toBeDefined();
		expect( component.getByText( 'This information will not be visible to others.' ) ).toBeDefined();
		expect( component ).toMatchSnapshot();
	} );

	describe( 'when the submit button is disabled', () => {
		it( 'renders the button disabled', () => {
			setUp( { isSubmitButtonDisabled: true } );
			expect( component.getByText( submitButtonTitle ) ).toBeDisabled();
		} );
	} );

	describe( 'when the form has validation errors', () => {
		it( 'shows the messages of the errors', async () => {
			const makeAndModelInput = component.queryByTestId( 'make-and-model-input' );

			await act( async () => {
				fireEvent( makeAndModelInput, 'change', '' );
			} );

			expect( component.queryByText( 'This field is mandatory.' ) ).toBeDefined();
		} );
	} );

	describe( 'when the submit button is pressed', () => {
		it( 'dismisses the keyboard and calls the form\'s onSubmit handler', async () => {
			const formOnSubmit = jest.spyOn( form, 'onSubmit' );
			const keyboardDismiss = jest.spyOn( Keyboard, 'dismiss' );
			form.onSubmit.mockImplementationOnce( () => {} );

			const submitButton = component.queryByTestId( 'submit-button' );
			fireEvent.press( submitButton );

			expect( keyboardDismiss ).toHaveBeenCalledTimes( 1 );
			expect( formOnSubmit ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
