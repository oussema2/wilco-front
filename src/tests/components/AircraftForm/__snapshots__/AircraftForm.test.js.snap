// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AircraftForm renders the AircraftForm screen correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        null,
      ]
    }
    testID="horizontalPadding-component"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "minHeight": 44,
        }
      }
      testID="navigationBar-component"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 44,
            },
            Object {
              "flex": 0.25,
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 15,
              "left": 15,
              "right": 15,
              "top": 15,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/icons/chevronLeft.png",
              }
            }
            style={
              Object {
                "height": 20,
                "width": 20,
              }
            }
            testID="backArrow-image"
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 44,
            },
            Object {
              "flex": 0.5,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          onPress={null}
          style={
            Object {
              "alignItems": "center",
              "color": "#000000",
              "flex": 1,
              "fontFamily": "NotoSansJP-Black",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": -0.3,
              "lineHeight": 22,
              "textAlign": "center",
            }
          }
          testID="title-text"
        >
          Do something with aircraft
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 44,
            },
            Object {
              "flex": 0.25,
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "width": 20,
            }
          }
          testID="rightPlaceholder-view"
        />
      </View>
    </View>
  </View>,
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
      style={Object {}}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "paddingHorizontal": 20,
              },
              null,
            ]
          }
          testID="horizontalPadding-component"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "column",
                "height": 165,
                "justifyContent": "space-between",
                "marginTop": 39,
              }
            }
            testID="edit-avatar"
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="edit-avatar-image-testID"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "aspectRatio": 1,
                    "backgroundColor": "#A0A0A0",
                    "borderRadius": 68,
                    "display": "flex",
                    "height": 132,
                    "justifyContent": "center",
                  }
                }
                testID="Avatar-Component"
              >
                <View
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Object {
                        "borderRadius": 68,
                        "height": "100%",
                        "width": "100%",
                      },
                    ]
                  }
                >
                  <FastImageView
                    onFastImageLoadEnd={[Function]}
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "avatar/source",
                      }
                    }
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                    testID="image"
                  />
                </View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "justifyContent": "center",
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "aspectRatio": 1,
                        "backgroundColor": "#A0A0A0",
                        "borderRadius": 68,
                        "display": "flex",
                        "height": 132,
                        "justifyContent": "center",
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "overflow": "hidden",
                          },
                          Object {
                            "height": 54,
                            "marginLeft": 6,
                            "width": 60,
                          },
                        ]
                      }
                    >
                      <FastImageView
                        resizeMode="cover"
                        source={
                          Object {
                            "testUri": "../../../src/assets/icons/aircraft.png",
                          }
                        }
                        style={
                          Object {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          }
                        }
                        testID="Avatar-Component-loader"
                        tintColor="#FFFFFF"
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "borderRadius": 3,
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 3,
                  },
                  undefined,
                  false,
                ]
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "busy": false,
                    "disabled": false,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="edit-avatar-button-testID"
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "flex-end",
                      "backgroundColor": "transparent",
                      "borderColor": "#2089dc",
                      "borderRadius": 26,
                      "borderWidth": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "padding": 0,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#12A3BE",
                        "fontFamily": "NotoSansJP-Bold",
                        "fontSize": 15,
                        "fontWeight": undefined,
                        "letterSpacing": 0,
                        "lineHeight": 20,
                        "paddingVertical": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    Change aircraft photo
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 24,
                "marginTop": 40,
                "minHeight": 98,
              }
            }
            testID="make-and-model-input"
          >
            <View
              style={
                Object {
                  "alignContent": "center",
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontFamily": "NotoSansJP-Black",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "letterSpacing": -0.3,
                    "lineHeight": 22,
                    "marginBottom": 4,
                  }
                }
                testID="input-label"
              >
                Make and model
              </Text>
              <Text
                style={
                  Object {
                    "color": "#D60A50",
                    "fontFamily": "NotoSansJP-Bold",
                    "fontSize": 14,
                    "fontWeight": undefined,
                    "letterSpacing": 0,
                    "lineHeight": 20,
                    "marginLeft": 4,
                  }
                }
                testID="input-required"
              >
                *
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
              >
                <View
                  collapsable={false}
                  nativeID="animatedComponent"
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#F2F2F2",
                      "borderBottomWidth": 0,
                      "borderColor": "#86939e",
                      "borderRadius": 8,
                      "flexDirection": "row",
                      "marginBottom": 5,
                      "minHeight": 52,
                      "paddingHorizontal": 12,
                      "textAlignVertical": "center",
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    autoCapitalize="words"
                    autoCorrect={true}
                    editable={true}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="e.g. Cirrus SR 22"
                    placeholderTextColor="#A0A0A0"
                    returnKeyType="default"
                    scrollEnabled={true}
                    secureTextEntry={false}
                    selectionColor="#12A3BE"
                    style={
                      Object {
                        "color": "#000000",
                        "flex": 1,
                        "fontFamily": "NotoSansJP-Bold",
                        "fontSize": 14,
                        "fontWeight": undefined,
                        "letterSpacing": 0,
                        "lineHeight": 20,
                        "minHeight": 40,
                        "paddingVertical": 0,
                      }
                    }
                    testID="input"
                    textAlignVertical="center"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "marginVertical": 4,
                        "paddingRight": 4,
                      }
                    }
                  />
                </View>
                <Text
                  style={
                    Object {
                      "color": "#D60A50",
                      "fontFamily": "NotoSansJP-Regular",
                      "fontSize": 12,
                      "fontWeight": undefined,
                      "letterSpacing": 0,
                      "lineHeight": 16,
                      "margin": 0,
                      "marginLeft": 0,
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 36,
                "minHeight": 98,
              }
            }
            testID="tail-number-input"
          >
            <View
              style={
                Object {
                  "alignContent": "center",
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontFamily": "NotoSansJP-Black",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "letterSpacing": -0.3,
                    "lineHeight": 22,
                    "marginBottom": 4,
                  }
                }
                testID="input-label"
              >
                Tail number
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
              >
                <View
                  collapsable={false}
                  nativeID="animatedComponent"
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#F2F2F2",
                      "borderBottomWidth": 0,
                      "borderColor": "#86939e",
                      "borderRadius": 8,
                      "flexDirection": "row",
                      "marginBottom": 5,
                      "minHeight": 52,
                      "paddingHorizontal": 12,
                      "textAlignVertical": "center",
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    autoCapitalize="characters"
                    autoCorrect={true}
                    editable={true}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="e.g. N2129J"
                    placeholderTextColor="#A0A0A0"
                    returnKeyType="default"
                    scrollEnabled={true}
                    secureTextEntry={false}
                    selectionColor="#12A3BE"
                    style={
                      Object {
                        "color": "#000000",
                        "flex": 1,
                        "fontFamily": "NotoSansJP-Bold",
                        "fontSize": 14,
                        "fontWeight": undefined,
                        "letterSpacing": 0,
                        "lineHeight": 20,
                        "minHeight": 40,
                        "paddingVertical": 0,
                      }
                    }
                    testID="input"
                    textAlignVertical="center"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "marginVertical": 4,
                        "paddingRight": 4,
                      }
                    }
                  />
                </View>
                <Text
                  style={
                    Object {
                      "color": "#D60A50",
                      "fontFamily": "NotoSansJP-Regular",
                      "fontSize": 12,
                      "fontWeight": undefined,
                      "letterSpacing": 0,
                      "lineHeight": 16,
                      "margin": 0,
                      "marginLeft": 0,
                    }
                  }
                >
                  
                </Text>
              </View>
              <Text
                style={
                  Object {
                    "color": "#000000",
                    "fontFamily": "NotoSansJP-Regular",
                    "fontSize": 12,
                    "fontWeight": undefined,
                    "letterSpacing": 0,
                    "lineHeight": 16,
                    "marginTop": -14,
                  }
                }
                testID="helperText-testID"
              >
                This information will not be visible to others.
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignSelf": "center",
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="submit-button"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#12A3BE",
                "borderColor": "#2089dc",
                "borderRadius": 26,
                "borderWidth": 0,
                "flexDirection": "row",
                "height": 52,
                "justifyContent": "center",
                "padding": 8,
                "width": 164,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "NotoSansJP-Bold",
                  "fontSize": 15,
                  "fontWeight": undefined,
                  "letterSpacing": 0.4,
                  "lineHeight": 20,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Submit
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;
